<?xml version="1.0"?>

<!-- Launch file for the starting mission for the thesis of Boris Dehem-->
<!-- Explore in a straight line until a target is found. Then explore around it, and return to base-->
<!-- Work In Progress -->
<launch>
  <!-- IPv4 address of your drones -->
  <arg name="drone_name" default="arnold" />
  <arg name="mb1_ip" default="192.168.1.1" />
  <!-- Ultrasound frequency (7 or 8). -->
  <arg name="freq" default="8" />


  <!-- [begin] RELAY TOPICS BETWEEN boris_drone_GUI AND ARDRONE_AUTONOMY -->
  <!-- if you need something more sophisticated, create your node and replace the corresponding relay -->
  <node name="relay_navdata" pkg="topic_tools" type="relay"  args="ardrone/navdata navdata"></node>
  <node name="relay_emergency" pkg="topic_tools" type="relay"  args="ardrone/reset"></node>
  <node name="relay_land" pkg="topic_tools" type="relay"  args="land ardrone/land"></node>
  <node name="relay_takeoff" pkg="topic_tools" type="relay"  args="takeoff ardrone/takeoff"></node>
  <!-- [end] RELAY TOPICS BETWEEN boris_drone_GUI AND ARDRONE_AUTONOMY -->

  <node name="ardrone_driver" pkg="ardrone_autonomy" type="ardrone_driver"
        output="screen" clear_params="true" args="-ip $(arg mb1_ip)">

    <param name="tf_prefix" value="$(arg drone_name)/autonomy_mb1" />
    <param name="outdoor" value="0" />
    <param name="looprate" value="30" />
    <!-- <param name="max_bitrate" value="4000" />
    <param name="bitrate" value="4000" />-->
    <param name="navdata_demo" value="0" />
    <param name="flight_without_shell" value="0" />
    <param name="altitude_max" value="4000" />
    <param name="altitude_min" value="50" />
    <param name="euler_angle_max" value="0.21" />
    <param name="control_vz_max" value="700" />
    <!-- <param name="control_yaw" value="1.75" /> -->
    <!-- <param name="detect_type" value="10" />
    <param name="enemy_colors" value="3" />
    <param name="detections_select_h" value="32" />
    <param name="detections_select_v_hsync" value="128" />
    <param name="enemy_without_shell" value="0" />-->
    <param name="ultrasound_freq" value="$(arg freq)" />
    <param name="realtime_navdata" value="true" />
    <param name="realtime_video" value="true" />
    <!-- Covariance Values (3x3 matrices reshaped to 1x9)-->
    <rosparam param="cov/imu_la">[0.1, 0.0, 0.0, 0.0, 0.1, 0.0, 0.0, 0.0, 0.1]</rosparam>
    <rosparam param="cov/imu_av">[1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]</rosparam>
    <rosparam param="cov/imu_or">[1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 100000.0]</rosparam>
  </node>

  <node name="image_proc_bottom" pkg="image_proc" type="image_proc" ns="ardrone/bottom"/>

  <node name="vision_gui" pkg="boris_drone" type="vision_gui" output="screen">
    <param name="draw_keypoints" value="true" />
    <param name="draw_target" value="true" />
  </node>

  <node name="boris_drone_pose_estimation" pkg="boris_drone" type="pose_estimation" output="screen">
    <param name="drone_prefix" value=""/>
    <param name="use_visual_pose" value="true" />
  </node>

  <!-- bottom camera -->
  <rosparam param="cam_matrix">[705.0, 0.0, 319.16, 0.0, 705.0, 219.6, 0.0, 0.0, 1.0]</rosparam>
  <rosparam param="img_size">[640.0, 360.0]</rosparam>
  <!-- bottom camera -->

  <!-- front camera -->
  <!-- <rosparam param="cam_matrix">[529.1, 0.0, 350.6, 0.0, 529.1, 182.2, 0.0, 0.0, 1.0]</rosparam>
  <rosparam param="img_size">[735.0, 360.0]</rosparam> -->
  <!-- front camera -->

  <node name="boris_drone_computer_vision" pkg="boris_drone" type="computer_vision" output="screen">
    <param name="drone_prefix" value=""/>
    <param name="video_channel" value="ardrone/bottom/image_rect_color"/>
    <param name="cam_type" value="bottom"/>
    <param name="use_OpticalFlowPyrLK" value="true"/>
  </node>

  <node name="boris_drone_map_keyframe_based" pkg="boris_drone" type="map_keyframe_based_1" output="screen">
    <param name="do_search" value="true" />
    <param name="stop_if_lost" value="true" />
  </node>

  <node name="boris_drone_path_planning" pkg="boris_drone" type="path_planning_1" output="screen">
    <param name="drone_prefix" value=""/>
  </node>

  <node name="boris_drone_controller" pkg="boris_drone" type="controller_custom" output="screen">
    <param name="drone_prefix" value=""/>
  </node>

  <node name="boris_drone_strategy" pkg="boris_drone" type="strategy_1" output="screen">
      <param name="drone_name" value="$(arg drone_name)" />
  </node>

</launch>
